dsl/grpc.go

```diff
@@ -0,0 +1,313 @@
+//         Field(1, "name", String, "Name of account")
+//         TokenField(2, "token", String, "JWT token for authentication")
+//     })
+//
+//     var CreateResult = ResultType("application/vnd.create", func() {
+//         Attributes(func() {
+//             Field(1, "name", String, "Name of the created resource")
+//             Field(2, "href", String, "Href of the created resource")
+//         })
+//     })
+//
+//     Method("create", func() {
+//         Payload(CreatePayload)
+//         Result(CreateResult)
+//         GRPC(func() {
+//             Message(func() {
+//                 Attribute("token") // "token" sent in the request message
+//                                    // along with "name"
+//             })
+//             Response(func() {
+//                 Code(CodeOK)
+//                 Message(func() {
+//                     Attribute("name") // "name" sent in the response
+//                                       // message along with "href"
+//                     Required("name")  // "name" is set to required
+//                 })
+//             })
+//         })
+//     })
+//
+// If the method Payload/Result type is a primitive, array, or a map the
+// request/response message by default contains one attribute with name
+// "field", "rpc:tag" set to 1, and the type set to the type of the
+// method Payload/Result. The function argument can also be used to set
+// the message field name to something other than "field".
+//
+// Example:
+//
+//     Method("add", func() {
+//         Payload(Operands)
+//         Result(Int)      // method Result is a primitive
+//         GRPC(func() {
+//             Response(CodeOK, func()
+//                 Message(func() {
+//                     Attribute("sum") // Response message has one field with
+//                                      // name "sum" instead of "field"
+//                 })
+//             })
+//         })
+//     })
+//
+func Message(fn func()) {
+	var setter func(*expr.AttributeExpr)
+	{
+		switch e := eval.Current().(type) {
+		case *expr.GRPCEndpointExpr:
+			setter = func(att *expr.AttributeExpr) {
+				e.Request = att
+			}
+		case *expr.GRPCErrorExpr:
+			setter = func(att *expr.AttributeExpr) {
+				if e.Response == nil {
+					e.Response = &expr.GRPCResponseExpr{}
+				}
+				e.Response.Message = att
+			}
+		case *expr.GRPCResponseExpr:
+			setter = func(att *expr.AttributeExpr) {
+				e.Message = att
+			}
+		default:
+			eval.IncompatibleDSL()
+			return
+		}
+	}
+	attr := &expr.AttributeExpr{}
+	if eval.Execute(fn, attr) {
+		setter(attr)
+	}
+}
+
+// Metadata defines a gRPC request metadata.
+//
+// Metadata must appear in a gRPC endpoint expression to describe gRPC request
+// metadata.
+//
+// Security attributes in the method Payload are automatically added to the
+// request metadata unless specified explicitly in request message using
+// Message DSL. All other attributes in method Payload are added to the
```

Comment:
using the Message function. I've tried to go away from using the term DSL in favor of just "function" when applicable. Might be good to review all uses of DSL and see if they can be replaced with "function".